File: i18n/en/docusaurus-plugin-content-docs/current/stack/3113_hard/max_boundary_subarrays.cpp
#include <stack>
#include <vector>
using namespace std;

long long count_subarrays(vector<int>& nums) { // LeetCode Q.3113.
    stack<pair<int, int>> stack; // Format: {num, count}.

    long long subarraysCount = 0;

    for (const auto& num : nums) {
        while (!stack.empty() && stack.top().first < num) stack.pop();

        if (!stack.empty() && stack.top().first == num)
            stack.top().second++;

        else
            stack.push({num, 1});

        subarraysCount += stack.top().second;
    }

    return subarraysCount;
}

File: docs/stack/3113_hard/max_boundary_subarrays.py

def count_subarrays(nums: list[int]) -> int:  # LeetCode Q.3113.
    stack: list[tuple[int, int]] = []  # Format: (num, count).

    subarrays_count = 0

    for num in nums:
        while stack and stack[-1][0] < num:
            stack.pop(-1)
        
        count = 1  # Base case: current num only.

        if stack and stack[-1][0] == num:
            count += stack.pop(-1)[1]

        subarrays_count += count
        stack.append((num, count))

    return subarrays_count
